h/streamer/websocket.py

```diff
@@ -229,3 +235,58 @@ def _expand_uris(session, clause):
         expanded.update(storage.expand_uri(session, item))
 
     clause["value"] = list(expanded)
+
+
+class NumericMetric:
+    """Helper for tracking changes in a numeric metric over time."""
+
+    def __init__(self, get_metric):
+        """
+        :param get_metric: Callable which returns the current value of the metric
+        """
+        self.get_metric = get_metric
+        self.prev_value = get_metric()
+
+    def delta(self):
+        """Update the metric and return the difference between current and previous values"""
+        new_value = self.get_metric()
+        delta = new_value - self.prev_value
+        self.prev_value = new_value
+        return delta
+
+
+@newrelic.agent.data_source_factory(name="WebSocket Connections")
+def websocket_metrics(settings, environ):
+    """
+    A New Relic metric data source which provides metrics about WebSocket
+    connections.
+
+    See https://docs.newrelic.com/docs/agents/python-agent/supported-features/python-custom-metrics.
+    """
+
+    socket_list = environ.get("socket_list", WebSocket.instances)  # Test seam
+
+    messages_sent_metric = NumericMetric(
+        lambda: sum([ws.messages_sent for ws in socket_list])
+    )
+    messages_received_metric = NumericMetric(
+        lambda: sum([ws.messages_received for ws in socket_list])
+    )
+    prefix = "Custom/WebSocket"
+
+    def generate_metrics():
+        active_connections = len(socket_list)
+        authenticated_connections = len(
+            [ws for ws in socket_list if ws.authenticated_userid]
+        )
+
+        yield (f"{prefix}/ActiveConnections", active_connections)
+        yield (f"{prefix}/AuthenticatedConnections", authenticated_connections)
+        yield (
+            f"{prefix}/AnonymousConnections",
+            active_connections - authenticated_connections,
+        )
```

Comment:
We could do with monitoring the queue size, as this causes us to stop accepting messages when full:

`yield (f"{prefix}/WorkQueueSize", WORK_QUEUE.qsize())`